### PR TITLE
Branching tutorial

### PR DESCRIPTION
**Pull request by dchetelat**
Simplified implementation of the Gasse et al. (2019) learning to branch paper as a Ecole tutorial.

Biggest differences:
- No parallelism in sample collection, which makes it very slow but simple to follow.
- No PreNorm layers, I used LayerNorm instead.
- No shuffling in the training dataset (maybe we should add this back.)

I used the "Easy" combinatorial auctions from the paper, with 1000 samples. This is too little to achieve good performance against SCIP's default brancher, a good question is whether it would be worth it to increase that, at the cost of making the running of the tutorial very slow (hours). Alternatively, maybe we could make instances smaller, perhaps that lessens the dataset size required to achieve good performance.

There is no timings in the benchmarking right now as well, although considering the poor performance there is little point right now. But it might be useful to include later to showcase it.

### CODE REVIEW

**Comment by AntoinePrv** [comment]
Seems good to me. Here are some small typos, improvement ideas:
 - Intro says TF, but notebook seems to use Pytorch
 - It would be nice to have a cell/file that lists the dependencies and version needed (conda env file maybe).
 - In the evaluation, instead of manually summing `nb_nodes` and `time`, you could showcase the use of `ecole.reward.NNodes().cumsum()`

**Review by AntoinePrv — APPROVED** [review]

**Comment by dchetelat** [comment]
>     * It would be nice to have a cell/file that lists the dependencies and version needed (conda env file maybe).

We would have to move the tutorials in different folders then? Alternatively we can simply list the version numbers in the notebook to keep it simple.

**Comment by AntoinePrv** [comment]
> Alternatively we can simply list the version numbers in the notebook to keep it simple.

Sounds good

**Comment by dchetelat** [comment]
Okay I fixed the text, added the dependencies in the text and added a final example of generalization to harder instances that also showcases the cumsum technique.

**Comment by gasse** [comment]
A few comments:
 - there is a typo: "showcasees"
 - the code could have more comments. I know it will mostly be redundant with the text, but it is better to repeat I think. Some people might only look at the code. I think of the `ExploreThenStrongBranch` class, in extract: "extract randomly either strong branching scores (expensive expert) or pseudocost scores (weak expert for exploration)". But mostly the data acquisition loop could use more comments.
 - it hinders readability to have 4 lines for basically transforming the `node_observation` before recording samples to disk. Maybe we could record the `node_observations` as is, and do the transformations later on, since those are specific to our model ?

**Comment by gasse** [comment]
Continued:
 - the PATIENCE and EARLY_STOPPING functionalities are useful to make things work in practice, but we don't really need those for the tutorial. Removing them would simplify the code. We can just train for X epochs and that's it, old school way.

**Comment by gasse** [comment]
Continued:
 - in the evaluation code, the time and nnodes from reset is not counted. A simple way to do that is to use cumsum() and report only the last value, when done==True.

**Comment by gasse** [comment]
Continued:
 - in the evaluation logs, I would maybe have the number of nodes of default SCIP / the learned model on top of each other, in order to compare more easily at a glance:
```
Instance   0 | SCIP nb nodes  956 | SCIP time    10.27
             | GNN  nb nodes 1294 | GNN  time     6.40
             | Gain       -35.36% | Gain        37.68%
```
versus
```
Instance   0 | SCIP nb nodes  956 | GNN nb nodes 1294 | Gain   -35.36%
             | SCIP time    10.27 | GNN time     6.40 | Gain    37.68%
```

**Comment by dchetelat** [comment]
Hi Maxime,


>     * it hinders readability to have 4 lines for basically transforming the `node_observation` before recording samples to disk. Maybe we could record the `node_observations` as is, and do the transformations later on, since those are specific to our model ?

It's like that because we cannot pickle the `ecole.core.observation.NodeBipartiteObs` object directly, so we have to unload it into a tuple of numpy arrays... Otherwise I agree, it would have been cleaner.

I followed the rest of your suggestions. I rebased also.

**Comment by AntoinePrv** [comment]
> It's like that because we cannot pickle the ecole.core.observation.NodeBipartiteObs object directly, so we have to unload it into a tuple of numpy arrays... Otherwise I agree, it would have been cleaner.

We can add pickle support for `NodeBipartiteObs`. If it is supported on `xtensor`'s end it is not much for us to add.  #106

@dchetelat for the tutorial, would it make sense to not pickle the data, and keep it in memory in a `list` of so?
